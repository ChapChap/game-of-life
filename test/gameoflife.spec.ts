import {
  getNeighbors,
  IContext,
  nextGeneration,
  render,
  stringToCoord
} from "../src/gameoflife";

describe("gameoflife", () => {
  describe("nextGeneration", () => {
    it("should kill a live cell with less than two neighbors", () => {
      const newGrid = nextGeneration(["1,1", "1,2"]);

      expect(newGrid).toEqual([]);
    });

    it("should let a live cell with 2 neighbors alive", () => {
      const cells = ["1,1", "1,2", "2,1"];
      const newGrid = nextGeneration(cells);

      const expected = cells.every((cell) => newGrid.includes(cell));

      expect(expected).toBeTruthy();
    });

    it("should let a live cell with 3 neighbors alive", () => {
      const cells = ["1,1", "1,2", "2,1", "2,2"];
      const newGrid = nextGeneration(cells);

      const expected = cells.every((cell) => newGrid.includes(cell));

      expect(expected).toBeTruthy();
    });

    it("should resurrect a dead cell with 3 live neighbors", () => {
      const newGrid = nextGeneration(["0,0", "1,0", "2,0"]);

      const expected = ["1,1", "1,-1"]
        .every((cell) => newGrid.includes(cell));

      expect(expected).toBeTruthy();
    });

    it("should kill a live cell with more than tree live neighbors", () => {
      const newGrid = nextGeneration(["0,0", "1,0", "2,0", "0,1", "1,1"]);

      const oneOne = newGrid.find((cell) => cell === "1,1");

      expect(oneOne).toBeFalsy();
    });
  });

  describe("getNeighbors", () => {
    it("should return all neighbors of a cell", () => {
      const neighbors = getNeighbors(1, 1);

      expect(neighbors).toEqual([
        "1,2",
        "2,2",
        "2,1",
        "2,0",
        "1,0",
        "0,0",
        "0,1",
        "0,2"
      ]);
    });
  });

  describe("stringToCoord", () => {
    it("should convert 2,1 into [2, 1]", () => {
      expect(stringToCoord("2,1")).toEqual([2, 1]);
    });
  });

  describe("render", () => {
    it("should render living cells on the context", () => {
      const cells = ["1,1", "1,2", "2,1"];

      const context: IContext = {
        canvas: { width: 100, height: 100 },
        fillRect: jest.fn(),
        fillStyle: ""
      };

      render(context, cells, 2);

      expect(context.fillRect).toMatchSnapshot();
    });
  });
});
