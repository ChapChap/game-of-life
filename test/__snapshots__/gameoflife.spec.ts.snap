// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gameoflife render should render living cells on the context 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      0,
      0,
      100,
      100,
    ],
    Array [
      10,
      10,
      10,
      10,
    ],
    Array [
      10,
      20,
      10,
      10,
    ],
    Array [
      20,
      10,
      10,
      10,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
